drivers/usb/device/usb_dc_numaker.c

```diff
@@ -0,0 +1,1943 @@
+	uint32_t clk_modidx;
+	uint32_t clk_src;
+	uint32_t clk_div;
+	const struct device *clkctrl_dev;
+	void (*irq_config_func)(struct usb_dc_numaker_device *dev);
+	void (*irq_unconfig_func)(struct usb_dc_numaker_device *dev);
+	const struct pinctrl_dev_config *pincfg;
+	uint32_t num_bidir_endpoints;
+	uint32_t dmabuf_size;
+	int maximum_speed;
+};
+
+/* EP context */
+struct numaker_usbd_ep {
+	bool valid; /* Valid */
+
+	bool nak_clr; /* NAK cleared (ACK next transaction) */
+
+	struct usb_dc_numaker_device *dev; /* Pointer to the containing device */
+
+	uint8_t usbd_hw_ep_hndl; /* BSP USBD driver EP handle EP0, EP1, EP2, etc */
+	uint32_t usbd_hw_ep_cfg; /* BSP USBD driver EP configuration */
+
+	/* EP DMA buffer */
+	bool dmabuf_valid;
+	uint32_t dmabuf_base;
+	uint32_t dmabuf_size;
+
+	/* On USBD, no H/W FIFO. Simulate based on above DMA buffer with
+	 * one-shot implementation
+	 */
+	uint32_t read_fifo_pos;
+	uint32_t read_fifo_used;
+	uint32_t write_fifo_pos;
+	uint32_t write_fifo_free;
+
+	/* NOTE: On USBD, Setup and CTRL OUT are not completely separated. CTRL OUT MXPLD
+	 * can be overridden to 8 by next Setup. To overcome it, we make one copy of CTRL
+	 * OUT MXPLD immediately on its interrupt.
+	 */
+	uint32_t mxpld_ctrlout;
+
+	/* EP address */
+	bool ep_addr_valid;
+	uint8_t ep_addr;
+
+	/* EP MPS */
+	bool ep_mps_valid;
+	uint16_t ep_mps;
+
+	usb_dc_ep_callback cb; /* EP callback function */
+};
+
+/* EP context manager */
+struct numaker_usbd_ep_mgmt {
+	/* EP context management
+	 *
+	 * Allocate-only, and de-allocate all on re-initialize in usb_dc_attach().
+	 */
+	uint8_t ep_idx;
+
+	/* DMA buffer management
+	 *
+	 * Allocate-only, and de-allocate all on re-initialize in usb_dc_attach().
+	 */
+	uint32_t dmabuf_pos;
+
+	/* Pass Setup packet from ISR to thread */
+	bool new_setup;
+	struct usb_setup_packet setup_packet;
+
+	struct numaker_usbd_ep ep_pool[NUMAKER_USBD_EP_MAXNUM];
+};
+
+/* Mutable device context */
+struct usb_dc_numaker_data {
+	uint8_t addr; /* Host assigned USB device address */
+
+	struct k_mutex sync_mutex;
+
+	/* Enable interrupt top/bottom halves processing
+	 *
+	 * Registered callbacks may use mutex or other kernel functions which are not supported
+	 * in interrupt context
+	 */
+	struct k_msgq msgq;
+	struct numaker_usbd_msg msgq_buf[CONFIG_USB_DC_NUMAKER_MSG_QUEUE_SIZE];
+
+	K_KERNEL_STACK_MEMBER(msg_hdlr_thread_stack,
+			      CONFIG_USB_DC_NUMAKER_MSG_HANDLER_THREAD_STACK_SIZE);
+	struct k_thread msg_hdlr_thread;
+
+	usb_dc_status_callback status_cb; /* Status callback function */
+
+	struct numaker_usbd_ep_mgmt ep_mgmt; /* EP management */
+};
+
+/* Device context */
+struct usb_dc_numaker_device {
+	const struct usb_dc_numaker_config *config;
+	struct usb_dc_numaker_data *data;
+};
+
+/* Function declaration */
+static inline struct usb_dc_numaker_device *usb_dc_numaker_device_get(void);
+
+/* Internal functions */
+
+static inline void numaker_usbd_lock(const struct usb_dc_numaker_device *dev)
```

Comment:
`void foobar(const struct usb_dc_numaker_device *dev)`
That is bad pattern, please just use device model, instantiate a device using devicetree and change parameters to 
`void foobar(const struct device *dev)`

Reply:
followed device driver model. changed to `struct device`